design/20210324.certificate-request-approve-deny.md

```diff
@@ -0,0 +1,191 @@
+
+### Verbs
+The verb must *always* be `approve`. This verb grants the approver the
+permissions to set *both* Approved and Denied conditions.
+
+### Resource Names
+The resource name is made up of 3 parts: the issuer kind, issuer group, and the
+issuer name.
+
+The issuer kind must be lower case, i.e. "ClusterIssuer" would become
+"clusterissuer".
+
+The issuer group should be the same string of the group of issuer.
+
+The name may be either a wild card '*' denoting all issuer names, or instead a
+string which must match the issuer name as it appears in the IssuerRef.
+
+An example of signing all `MyIssuer` issuers, and `ClusterMyIssuers` with the name
+`MyApp`, in the `my-example.io` group:
+
+```yaml
+    resourceNames: ["myissuer.my-example.io/*", "clustermyissuer.my-example.io/MyApp"]
+```
+
+## cert-manager Approver
+
+The cert-manager controller is deployed with an internal approver controller.
+This controller will attempt to Approve *all* CertificateRequests that are
+created, regardless of its contents or IssuerRef.
+
+By default, this approver is given the permission to sign all internal issuers,
+that is `Issuer` and `ClusterIssuer` `cert-manager.io` issuers. External issuers
+may consider whether they wish to install further RBAC that would allow the
+cert-manager approver controller to also approve CertificateRequests that
+reference these external issuers.
+
+The default cert-manager approver controller may be disabled by adding the
+`--controllers=-certificaterequests-approver` argument to the
+`cert-manager-container` component. This allows for other approvers to make
+decisions about CertificateRequests, without racing against the internal
+approver controller.
+
+
+## Certificates Controller Changes
+
+### Issuing Controller
+
+Since external issuers have been built before the addition of Approved and
+Denied conditions, the issuing controller needs to be permissive. An external
+issuer may not honour an Approved condition and will sign and set a
+CertificateRequest as being Ready, before the request has been approved. The
+issuing controller must mark issuance as being successful in this case. In
+practice, this means that the issuing controller is never concerned with
+Approved conditions.
+
+External issuers that do not honour Denied conditions will sign
+CertificateRequests, even if they have a Denied condition set. In this case, the
+issuing controller will successfully complete the issuance of the Certificate.
+
+External Issuers and internal issuer that honour the Denied condition will never
+sign CertificateRequests with the Denied condition set, and thus never set the
+Ready condition. In this case, the issuing controller will consider this
+CertificateRequest as failed, set the Issuing condition to False, and set the
+last failure time. This will cause the issuance to be retried at a later date (1
+hour). This is done for a number of reasons:
```

Comment:
Do we want it to be retried if it's being denied..? Not totally against this, but if we add it now then we have to stick with this behavior is all

Reply:
As an end user, I do _feel_ like this it the correct way forward. Some policy decisions are made using external information which we have no way to know about (only sign certificates on a Tuesday kind of thing). This seems like the safest approach to me.

My only concern about this is the addition of the [`revisonHistoryLimit`](https://github.com/jetstack/cert-manager/pull/3773). In it's current form which I think is correct, it will only clean up CRs if the Certificate itself is ready. This means a CR which is constantly denied will produce a growing number of CRs. I think this is fine.. this is currently the existing behaviour anyway.